src/coreComponents/functions/MultivariableTableFunctionKernels.hpp

```diff
@@ -0,0 +1,327 @@
+    real64 axisLows[numDims];
+    real64 axisMults[numDims];
+
+    for( int i = 0; i < numDims; ++i )
+    {
+      integer const axisIndex = getAxisIntervalIndexLowMult( coordinates[i],
+                                                             m_axisMinimums[i], m_axisMaximums[i],
+                                                             m_axisSteps[i], m_axisStepInvs[i], m_axisPoints[i],
+                                                             axisLows[i], axisMults[i] );
+      hypercubeIndex += axisIndex * m_axisHypercubeMults[i];
+    }
+
+    interpolatePointWithDerivatives( coordinates,
+                                     getHypercubeData( hypercubeIndex ),
+                                     &axisLows[0], &axisMults[0],
+                                     &m_axisStepInvs[0],
+                                     values,
+                                     derivatives );
+
+  }
+
+protected:
+
+  /**
+   * @brief Get pointer to hypercube data
+   *
+   * @param[in] hypercubeIndex
+   * @return pointer to hypercube data
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  real64 const *
+  getHypercubeData( globalIndex const hypercubeIndex ) const
+  {
+    return &m_hypercubeData[hypercubeIndex * numVerts * numOps];
+  }
+
+  /**
+   * @brief Get the interval index, low and mult values for a given axis coordinate
+   *
+   * @param[in] axisCoordinate coordinate on a given axis
+   * @param[in] axisMin minimum value on a given axis
+   * @param[in] axisMax maximum value on a given axis
+   * @param[in] axisStep interval length for a given axis
+   * @param[in] axisStepInv inversion of the interval length for a given axis
+   * @param[in] axisPoints number of discretization points for a given axis
+   * @param[out] axisLow left coordinate of target axis interval
+   * @param[out] axisMult weight of the right coordinate of target axis interval
+   * @return integer target axis interval index
+   */
+  GEOSX_HOST_DEVICE
+  inline
+  integer
+  getAxisIntervalIndexLowMult( real64 const axisCoordinate,
+                               real64 const axisMin,
+                               real64 const axisMax,
+                               real64 const axisStep,
+                               real64 const axisStepInv,
+                               integer const axisPoints,
+                               real64 & axisLow,
+                               real64 & axisMult ) const
+  {
+    integer axisIntervalIndex = integer((axisCoordinate - axisMin) * axisStepInv );
+
+    // check that axisindex is within interpolation interval: valid axisindex is between 0 and (axisPoints - 2),
+    // since there are axisPoints-1 intervals along the axis
+
+    if( axisIntervalIndex < 0 )
+    {
+      axisIntervalIndex = 0;
+      if( axisCoordinate < axisMin )
+      {
+        GEOSX_WARNING( GEOSX_FMT( "Axis coordinate is out of limits ({}; {}) with value {}, extrapolation is applied", axisMin, axisMax, axisCoordinate ));
```

Comment:
You may have to change it back to `printf` unfortunately. `GEOSX_FMT` cannot be used inside a kernel as it is calling `{fmt}` which is not a gpu-compatible library and the call will fail to compile with nvcc. For the moment, `printf` is the only portable way to print from kernels.

Reply:
Ah, ok, didn`t think about that. Thank you!